File: apps/mobile/lib/services/android_auto_navigation_projection.dart
import '../domain/distance_unit.dart';
import '../domain/imported_route.dart';
import 'guidance_time_remaining.dart';
import 'navigation_guidance.dart';
import 'route_journey_progress.dart';
import 'route_progress.dart';

enum AndroidAutoNavigationHostEventType {
  started,
  stopped,
  externalDestination,
  autoDriveEnabled,
  rerouteRequested,
  arrived,
  restorationAcknowledged,
}

/// A bounded event sent from the Android Auto host back to the Dart owner.
class AndroidAutoNavigationHostEvent {
  const AndroidAutoNavigationHostEvent({
    required this.type,
    this.navigationSessionId,
    this.routeId,
    this.destination,
    this.reason,
    this.projectionSequence,
  });

  final AndroidAutoNavigationHostEventType type;
  final String? navigationSessionId;
  final String? routeId;
  final GeoPoint? destination;
  final String? reason;
  final int? projectionSequence;

  static AndroidAutoNavigationHostEvent? tryParse(Object? raw) {
    if (raw is! Map) return null;
    final rawType = raw['type'];
    final type = rawType is String
        ? AndroidAutoNavigationHostEventType.values
              .where((candidate) => candidate.name == rawType)
              .firstOrNull
        : null;
    if (type == null) return null;
    final navigationSessionId = _boundedOptionalString(
      raw['navigationSessionId'],
      200,
    );
    final routeId = _boundedOptionalString(raw['routeId'], 160);
    final reason = _boundedOptionalString(raw['reason'], 160);
    if ((raw['navigationSessionId'] != null && navigationSessionId == null) ||
        (raw['routeId'] != null && routeId == null) ||
        (raw['reason'] != null && reason == null)) {
      return null;
    }
    final sequence = raw['projectionSequence'];
    if (sequence != null && (sequence is! int || sequence < 1)) return null;
    final destination = _point(raw['destination']);
    if (raw['destination'] != null && destination == null) return null;
    return AndroidAutoNavigationHostEvent(
      type: type,
      navigationSessionId: navigationSessionId,
      routeId: routeId,
      destination: destination,
      reason: reason,
      projectionSequence: sequence as int?,
    );
  }

  static String? _boundedOptionalString(Object? raw, int maximumLength) {
    if (raw == null) return null;
    if (raw is! String) return null;
    final trimmed = raw.trim();
    if (trimmed.isEmpty) return null;
    return trimmed.length <= maximumLength
        ? trimmed
        : trimmed.substring(0, maximumLength);
  }

  static GeoPoint? _point(Object? raw) {
    if (raw == null) return null;
    if (raw is! Map) return null;
    final latitude = raw['latitude'];
    final longitude = raw['longitude'];
    if (latitude is! num || longitude is! num) return null;
    final lat = latitude.toDouble();
    final lon = longitude.toDouble();
    if (!lat.isFinite ||
        !lon.isFinite ||
        lat < -90 ||
        lat > 90 ||
        lon < -180 ||
        lon > 180) {
      return null;
    }
    return GeoPoint(latitude: lat, longitude: lon);
  }
}

/// Versioned navigation state intended only for the native Android Auto adapter.
///
/// The legacy top-level snapshot remains present while the Android car surface
/// migrates. Keeping this payload separate from CarPlay lets each native host
/// evolve its lifecycle without making the other platform guess at semantics.
Map<String, Object?> projectAndroidAutoNavigationV2({
  required String sourceId,
  required int sequence,
  required DateTime generatedAt,
  required String ridePhase,
  required ImportedRoute? route,
  required bool navigationEnabled,
  required NavigationGuidance? guidance,
  required DistanceUnit? distanceUnit,
  required String? localeIdentifier,
  required double? speedMetersPerSecond,
  required RouteJourneyProgress? journeyProgress,
  required RouteProgressGeometry? routeProgress,
  required bool followRider,
  required bool canPlanRoute,
  required bool canFreeRoam,
  required bool canStartPreparedRide,
  required Map<String, Object?>? alert,
}) {
  final navigationPhase = switch ((route, ridePhase, navigationEnabled)) {
    (null, _, _) => 'inactive',
    (_, 'endedRide', _) => 'ended',
    (_, _, false) => 'ended',
    (_, 'activeRide', true) => 'navigating',
    _ => 'routeReady',
  };
  final trafficSide = _trafficSide(
    guidance?.instruction,
    route?.maneuvers.firstOrNull,
  );
  final navigating = navigationPhase == 'navigating';

  return {
    'schemaVersion': 2,
    'sourceId': sourceId,
    'sequence': sequence,
    'generatedAtMillis': generatedAt.millisecondsSinceEpoch,
    'rideLifecycle': {'phase': ridePhase},
    'navigationLifecycle': {
      'phase': navigationPhase,
      // This is a desired host state, not the ride lifecycle. Android may lose
      // navigation ownership while the phone continues recording the ride.
      'shouldOwnNavigation': navigating,
    },
    'route': route == null
        ? null
        : {
            'id': route.id,
            'navigationSessionId': '${route.id}:android-navigation',
            'restorationId': route.id,
            'name': route.name,
            'trafficSide': trafficSide,
          },
    'currentManeuver': guidance == null
        ? null
        : _maneuverSnapshot(
            maneuver: guidance.maneuver,
            instruction: guidance.instruction,
            distanceMeters: guidance.distanceMeters,
            secondsRemaining: guidanceSecondsRemaining(
              distanceMeters: guidance.distanceMeters,
              speedMetersPerSecond: speedMetersPerSecond,
            ),
          ),
    'followingManeuver': switch ((
      guidance?.followingManeuver,
      guidance?.followingInstruction,
    )) {
      (final maneuver?, final instruction?) => _maneuverSnapshot(
        maneuver: maneuver,
        instruction: instruction,
        distanceMeters: guidance?.followingDistanceMeters,
      ),
      _ => null,
    },
    'journey': journeyProgress?.toSnapshot(),
    'progress': {
      'travelledMeters': _nonNegativeFinite(routeProgress?.progressMeters),
      'totalMeters': _nonNegativeFinite(routeProgress?.totalMeters),
    },
    'units': {
      'distance': distanceUnit?.name,
      'speed': switch (distanceUnit) {
        DistanceUnit.miles => 'milesPerHour',
        DistanceUnit.kilometres => 'kilometresPerHour',
        null => null,
      },
    },
    'localeIdentifier': localeIdentifier,
    'camera': {'followRider': followRider},
    'actions': {
      'canPlanRoute': canPlanRoute,
      'canFreeRoam': canFreeRoam,
      'canStartPreparedRide': canStartPreparedRide,
      'canCancelNavigation': navigating,
      'canLeaveRide': ridePhase == 'activeRide',
    },
    'alert': alert,
  };
}

Map<String, Object?> _maneuverSnapshot({
  required RouteManeuver maneuver,
  required ManeuverInstruction instruction,
  required double? distanceMeters,
  double? secondsRemaining,
}) => {
  'id': maneuver.identity,
  'kind': instruction.kind.name,
  'direction': instruction.direction.name,
  'engineType': maneuver.type,
  'engineModifier': maneuver.modifier,
  'instructionVariants': _variants([
    instruction.text,
    instruction.standaloneText,
  ]),
  'roadNameVariants': _variants([
    instruction.roadLabel,
    instruction.roadName,
    instruction.roadRef,
  ]),
  'position': {
    'latitude': maneuver.position.latitude,
    'longitude': maneuver.position.longitude,
  },
  'exitNumber': instruction.exitNumber ?? maneuver.exitNumber,
  'trafficSide': _trafficSide(instruction, maneuver),
  'distanceMeters': _nonNegativeFinite(distanceMeters),
  'secondsRemaining': _nonNegativeFinite(secondsRemaining),
  'bearingBeforeDegrees': _finite(maneuver.bearingBeforeDegrees),
  'bearingAfterDegrees': _finite(maneuver.bearingAfterDegrees),
  'departureBearingDegrees': _finite(instruction.departureBearingDegrees),
  'stepCount': instruction.stepCount,
  'lanes': [
    for (final lane in instruction.lanes)
      {'indications': lane.indications, 'valid': lane.valid},
  ],
};

List<String> _variants(Iterable<String?> candidates) {
  final variants = <String>[];
  for (final candidate in candidates) {
    final value = candidate?.trim();
    if (value == null || value.isEmpty || variants.contains(value)) continue;
    variants.add(value);
  }
  return variants;
}

String _trafficSide(ManeuverInstruction? instruction, RouteManeuver? maneuver) {
  final explicit = instruction?.leftHandTraffic;
  if (explicit != null) return explicit ? 'left' : 'right';
  return switch (maneuver?.drivingSide?.trim().toLowerCase()) {
    'left' => 'left',
    'right' => 'right',
    _ => 'unknown',
  };
}

double? _nonNegativeFinite(double? value) {
  final finite = _finite(value);
  return finite != null && finite >= 0 ? finite : null;
}

double? _finite(double? value) => value?.isFinite == true ? value : null;

File: apps/mobile/android/app/src/main/kotlin/me/osholt/ride_relay/AndroidAutoNavigationProjection.kt
package me.osholt.ride_relay

/** Strict decoder for the versioned Android Auto navigation projection. */
internal data class AndroidAutoNavigationProjectionV2(
    val sourceId: String,
    val sequence: Long,
    val generatedAtMillis: Long,
    val ridePhase: String,
    val navigationPhase: String,
    val shouldOwnNavigation: Boolean,
    val route: Route?,
    val currentManeuver: Maneuver?,
    val followingManeuver: Maneuver?,
    val journey: Journey?,
    val progress: Progress,
    val units: Units,
    val localeIdentifier: String?,
    val followRider: Boolean,
    val actions: Actions,
    val alert: Alert?,
) {
    internal data class Route(
        val id: String,
        val navigationSessionId: String,
        val restorationId: String,
        val name: String,
        val trafficSide: String,
    )

    internal data class Position(val latitude: Double, val longitude: Double)

    internal data class Lane(val indications: List<String>, val valid: Boolean)

    internal data class Maneuver(
        val id: String,
        val kind: String,
        val direction: String,
        val engineType: String,
        val engineModifier: String?,
        val instructionVariants: List<String>,
        val roadNameVariants: List<String>,
        val position: Position,
        val exitNumber: Int?,
        val trafficSide: String,
        val distanceMeters: Double?,
        val secondsRemaining: Double?,
        val bearingBeforeDegrees: Double?,
        val bearingAfterDegrees: Double?,
        val departureBearingDegrees: Double?,
        val stepCount: Int,
        val lanes: List<Lane>,
    )

    internal data class Journey(
        val remainingDistanceMeters: Double?,
        val remainingSeconds: Long?,
        val arrivalTimeMillis: Long?,
        val nextWaypointName: String?,
        val nextWaypointDistanceMeters: Double?,
        val nextWaypointArrivalTimeMillis: Long?,
    )

    internal data class Progress(val travelledMeters: Double?, val totalMeters: Double?)

    internal data class Units(val distance: String?, val speed: String?)

    internal data class Actions(
        val canPlanRoute: Boolean,
        val canFreeRoam: Boolean,
        val canStartPreparedRide: Boolean,
        val canCancelNavigation: Boolean,
        val canLeaveRide: Boolean,
    )

    internal data class Alert(val message: String, val severity: String)

    companion object {
        private val ridePhases = setOf("home", "preRide", "activeRide", "endedRide")
        private val navigationPhases = setOf("inactive", "routeReady", "navigating", "ended")
        private val maneuverKinds = setOf(
            "depart",
            "arrive",
            "roundabout",
            "turn",
            "endOfRoad",
            "merge",
            "fork",
            "onRamp",
            "offRamp",
            "useLane",
            "continueAhead",
        )
        private val maneuverDirections = setOf(
            "sharpLeft",
            "left",
            "slightLeft",
            "straight",
            "slightRight",
            "right",
            "sharpRight",
            "uTurn",
            "unstated",
        )
        private val trafficSides = setOf("left", "right", "unknown")

        fun fromSnapshot(snapshot: Map<String, Any?>): AndroidAutoNavigationProjectionV2? {
            val raw = snapshot["androidAutoNavigation"] as? Map<*, *> ?: return null
            if (raw.integer("schemaVersion") != 2) return null
            val sourceId = raw.boundedString("sourceId", 96) ?: return null
            val sequence = raw.long("sequence")?.takeIf { it > 0 } ?: return null
            val generatedAtMillis = raw.long("generatedAtMillis")?.takeIf { it >= 0 } ?: return null
            val rideLifecycle = raw["rideLifecycle"] as? Map<*, *> ?: return null
            val ridePhase = rideLifecycle.boundedString("phase", 32)
                ?.takeIf(ridePhases::contains) ?: return null
            val navigationLifecycle = raw["navigationLifecycle"] as? Map<*, *> ?: return null
            val navigationPhase = navigationLifecycle.boundedString("phase", 32)
                ?.takeIf(navigationPhases::contains) ?: return null
            val shouldOwnNavigation = navigationLifecycle.boolean("shouldOwnNavigation")
                ?: return null
            if (shouldOwnNavigation != (navigationPhase == "navigating")) return null

            val route = when (val rawRoute = raw["route"]) {
                null -> null
                is Map<*, *> -> route(rawRoute) ?: return null
                else -> return null
            }
            if (navigationPhase != "inactive" && route == null) return null
            if (navigationPhase == "inactive" && route != null) return null

            val current = optionalMap(raw, "currentManeuver", ::maneuver) ?: return null
            val following = optionalMap(raw, "followingManeuver", ::maneuver) ?: return null
            val journey = optionalMap(raw, "journey", ::journey) ?: return null
            val progress = (raw["progress"] as? Map<*, *>)?.let(::progress) ?: return null
            val units = (raw["units"] as? Map<*, *>)?.let(::units) ?: return null
            val locale = raw.optionalBoundedString("localeIdentifier", 48) ?: return null
            val camera = raw["camera"] as? Map<*, *> ?: return null
            val followRider = camera.boolean("followRider") ?: return null
            val actions = (raw["actions"] as? Map<*, *>)?.let(::actions) ?: return null
            if (actions.canCancelNavigation != shouldOwnNavigation) return null
            val alert = optionalMap(raw, "alert", ::alert) ?: return null

            return AndroidAutoNavigationProjectionV2(
                sourceId = sourceId,
                sequence = sequence,
                generatedAtMillis = generatedAtMillis,
                ridePhase = ridePhase,
                navigationPhase = navigationPhase,
                shouldOwnNavigation = shouldOwnNavigation,
                route = route,
                currentManeuver = current.value,
                followingManeuver = following.value,
                journey = journey.value,
                progress = progress,
                units = units,
                localeIdentifier = locale.value,
                followRider = followRider,
                actions = actions,
                alert = alert.value,
            )
        }

        private fun route(raw: Map<*, *>): Route? {
            val trafficSide = raw.boundedString("trafficSide", 12)
                ?.takeIf(trafficSides::contains) ?: return null
            return Route(
                id = raw.boundedString("id", 160) ?: return null,
                navigationSessionId = raw.boundedString("navigationSessionId", 200) ?: return null,
                restorationId = raw.boundedString("restorationId", 200) ?: return null,
                name = raw.boundedString("name", 160) ?: return null,
                trafficSide = trafficSide,
            )
        }

        private fun maneuver(raw: Map<*, *>): Maneuver? {
            val kind = raw.boundedString("kind", 32)?.takeIf(maneuverKinds::contains)
                ?: return null
            val direction = raw.boundedString("direction", 32)
                ?.takeIf(maneuverDirections::contains) ?: return null
            val instructions = raw.stringList("instructionVariants", 4, 180)
                ?.takeIf(List<String>::isNotEmpty) ?: return null
            val roadNames = raw.stringList("roadNameVariants", 4, 160) ?: return null
            val position = (raw["position"] as? Map<*, *>)?.let(::position) ?: return null
            val trafficSide = raw.boundedString("trafficSide", 12)
                ?.takeIf(trafficSides::contains) ?: return null
            val exitNumber = raw.optionalInteger("exitNumber", 1..99) ?: return null
            val distance = raw.optionalNonNegativeDouble("distanceMeters") ?: return null
            val seconds = raw.optionalNonNegativeDouble("secondsRemaining") ?: return null
            val before = raw.optionalBearing("bearingBeforeDegrees") ?: return null
            val after = raw.optionalBearing("bearingAfterDegrees") ?: return null
            val departure = raw.optionalBearing("departureBearingDegrees") ?: return null
            val stepCount = raw.integer("stepCount")?.takeIf { it in 1..32 } ?: return null
            val lanes = lanes(raw["lanes"]) ?: return null
            val engineModifier = raw.optionalBoundedString("engineModifier", 80) ?: return null
            return Maneuver(
                id = raw.boundedString("id", 220) ?: return null,
                kind = kind,
                direction = direction,
                engineType = raw.boundedString("engineType", 80) ?: return null,
                engineModifier = engineModifier.value,
                instructionVariants = instructions,
                roadNameVariants = roadNames,
                position = position,
                exitNumber = exitNumber.value,
                trafficSide = trafficSide,
                distanceMeters = distance.value,
                secondsRemaining = seconds.value,
                bearingBeforeDegrees = before.value,
                bearingAfterDegrees = after.value,
                departureBearingDegrees = departure.value,
                stepCount = stepCount,
                lanes = lanes,
            )
        }

        private fun lanes(raw: Any?): List<Lane>? {
            val values = raw as? List<*> ?: return null
            if (values.size > 8) return null
            return values.map { value ->
                val lane = value as? Map<*, *> ?: return null
                Lane(
                    indications = lane.stringList("indications", 4, 32) ?: return null,
                    valid = lane.boolean("valid") ?: return null,
                )
            }
        }

        private fun position(raw: Map<*, *>): Position? {
            val latitude = raw.finiteDouble("latitude")?.takeIf { it in -90.0..90.0 }
                ?: return null
            val longitude = raw.finiteDouble("longitude")?.takeIf { it in -180.0..180.0 }
                ?: return null
            return Position(latitude, longitude)
        }

        private fun journey(raw: Map<*, *>): Journey? {
            val remainingDistance = raw.optionalNonNegativeDouble("remainingDistanceMeters")
                ?: return null
            val remainingSeconds = raw.optionalNonNegativeLong("remainingSeconds") ?: return null
            val arrival = raw.optionalNonNegativeLong("arrivalTimeMillis") ?: return null
            val waypointName = raw.optionalBoundedString("nextWaypointName", 160) ?: return null
            val waypointDistance = raw.optionalNonNegativeDouble("nextWaypointDistanceMeters")
                ?: return null
            val waypointArrival = raw.optionalNonNegativeLong("nextWaypointArrivalTimeMillis")
                ?: return null
            return Journey(
                remainingDistanceMeters = remainingDistance.value,
                remainingSeconds = remainingSeconds.value,
                arrivalTimeMillis = arrival.value,
                nextWaypointName = waypointName.value,
                nextWaypointDistanceMeters = waypointDistance.value,
                nextWaypointArrivalTimeMillis = waypointArrival.value,
            )
        }

        private fun progress(raw: Map<*, *>): Progress? {
            val travelled = raw.optionalNonNegativeDouble("travelledMeters") ?: return null
            val total = raw.optionalNonNegativeDouble("totalMeters") ?: return null
            if (travelled.value != null && total.value != null && travelled.value > total.value) {
                return null
            }
            return Progress(travelled.value, total.value)
        }

        private fun units(raw: Map<*, *>): Units? {
            val distance = raw.optionalBoundedString("distance", 24) ?: return null
            val speed = raw.optionalBoundedString("speed", 32) ?: return null
            if (distance.value !in setOf(null, "miles", "kilometres")) return null
            if (speed.value !in setOf(null, "milesPerHour", "kilometresPerHour")) return null
            if ((distance.value == "miles") != (speed.value == "milesPerHour")) return null
            if ((distance.value == "kilometres") != (speed.value == "kilometresPerHour")) return null
            return Units(distance.value, speed.value)
        }

        private fun actions(raw: Map<*, *>): Actions? = Actions(
            canPlanRoute = raw.boolean("canPlanRoute") ?: return null,
            canFreeRoam = raw.boolean("canFreeRoam") ?: return null,
            canStartPreparedRide = raw.boolean("canStartPreparedRide") ?: return null,
            canCancelNavigation = raw.boolean("canCancelNavigation") ?: return null,
            canLeaveRide = raw.boolean("canLeaveRide") ?: return null,
        )

        private fun alert(raw: Map<*, *>): Alert? = Alert(
            message = raw.boundedString("message", 180) ?: return null,
            severity = raw.boundedString("severity", 32) ?: return null,
        )

        /**
         * Distinguishes a valid explicit null from a malformed value. Platform
         * channels preserve nulls, and every optional field must be intentional.
         */
        private data class Optional<out T>(val value: T?)

        private fun <T> optionalMap(
            raw: Map<*, *>,
            key: String,
            decode: (Map<*, *>) -> T?,
        ): Optional<T>? = when (val value = raw[key]) {
            null -> Optional(null)
            is Map<*, *> -> decode(value)?.let(::Optional)
            else -> null
        }

        private fun Map<*, *>.optionalBoundedString(key: String, max: Int): Optional<String>? {
            val value = this[key] ?: return Optional(null)
            return (value as? String)?.trim()?.takeIf(String::isNotEmpty)
                ?.take(max)?.let(::Optional)
        }

        private fun Map<*, *>.optionalInteger(key: String, range: IntRange): Optional<Int>? {
            val value = this[key] ?: return Optional(null)
            return (value as? Number)?.toInt()?.takeIf(range::contains)?.let(::Optional)
        }

        private fun Map<*, *>.optionalNonNegativeDouble(key: String): Optional<Double>? {
            val value = this[key] ?: return Optional(null)
            return (value as? Number)?.toDouble()?.takeIf { it.isFinite() && it >= 0 }
                ?.let(::Optional)
        }

        private fun Map<*, *>.optionalNonNegativeLong(key: String): Optional<Long>? {
            val value = this[key] ?: return Optional(null)
            return (value as? Number)?.toLong()?.takeIf { it >= 0 }?.let(::Optional)
        }

        private fun Map<*, *>.optionalBearing(key: String): Optional<Double>? {
            val value = this[key] ?: return Optional(null)
            return (value as? Number)?.toDouble()?.takeIf { it.isFinite() && it in 0.0..360.0 }
                ?.let(::Optional)
        }

        private fun Map<*, *>.boundedString(key: String, max: Int): String? =
            (this[key] as? String)?.trim()?.takeIf(String::isNotEmpty)?.take(max)

        private fun Map<*, *>.stringList(
            key: String,
            maxCount: Int,
            maxLength: Int,
        ): List<String>? {
            val values = this[key] as? List<*> ?: return null
            if (values.size > maxCount) return null
            val result = mutableListOf<String>()
            values.forEach { value ->
                val text = (value as? String)?.trim()?.takeIf(String::isNotEmpty)
                    ?.take(maxLength) ?: return null
                if (text !in result) result += text
            }
            return result
        }

        private fun Map<*, *>.integer(key: String): Int? = (this[key] as? Number)?.toInt()

        private fun Map<*, *>.long(key: String): Long? = (this[key] as? Number)?.toLong()

        private fun Map<*, *>.boolean(key: String): Boolean? = this[key] as? Boolean

        private fun Map<*, *>.finiteDouble(key: String): Double? =
            (this[key] as? Number)?.toDouble()?.takeIf(Double::isFinite)
    }
}

/**
 * Orders projections independently from ride state.
 *
 * A new Flutter bridge restarts its sequence at one, so a newer source may take
 * ownership by generation time. Replayed commands from the current source and
 * older commands from a retired source are rejected before they can revive
 * ended guidance or duplicate a native navigation session.
 */
internal class AndroidAutoNavigationProjectionStore {
    var latest: AndroidAutoNavigationProjectionV2? = null
        private set

    fun accept(snapshot: Map<String, Any?>): Boolean {
        val candidate = AndroidAutoNavigationProjectionV2.fromSnapshot(snapshot) ?: return false
        latest?.let { current ->
            if (candidate.sourceId == current.sourceId) {
                if (candidate.sequence <= current.sequence) return false
            } else if (candidate.generatedAtMillis < current.generatedAtMillis) {
                return false
            }
        }
        latest = candidate
        return true
    }

    fun clear() {
        latest = null
    }
}
